### PR TITLE
inductor: support conv+binary foldinig for freezing path

### PR DESCRIPTION
**Pull request by XiaobingSuper**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105651
* #105650
* __->__ #105048

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/fx_passes/freezing_patterns.py [hunk 1]
```diff
@@ -118,3 +124,132 @@ def addmm_fuse_replacement_second(inp, w1, w2, w3, b1, b2, b3):
...
+    _computation_ops = [aten.convolution.default]
+    _binary_ops = [aten.add.Tensor, aten.sub.Tensor, aten.mul.Tensor, aten.div.Tensor]
+    _computation_calls = [CallFunction(aten.convolution.default, *_conv_args)]
+
+    def _check_conv_and_broadcast_op(conv_node, other):
+        if not (isinstance(other, torch.fx.Node) and other.op != "get_attr"):
+            return False
+        if conv_node.args[1].op != "get_attr" or (
+            conv_node.args[2] is not None and conv_node.args[2].op != "get_attr"
+        ):
```
Comment (jgong5) [inline torch/_inductor/fx_passes/freezing_patterns.py:141]: I feel it is more readable to factor out a function, something like `is_constant(conv_node.args[1])`  to check whether some input is constant.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/fx_passes/freezing_patterns.py [hunk 2]
```diff
@@ -43,12 +47,14 @@ def freezing_passes(gm: torch.fx.GraphModule):
 
 @init_once_fakemode
 def lazy_init():
+    """
     if torch._C._has_mkldnn and config.cpp.weight_prepack:
         from .mkldnn_fusion import _mkldnn_weight_pack_init
 
         _mkldnn_weight_pack_init()
-
+    """
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply torch/_inductor/fx_passes/freezing_patterns.py]: it is just for debugging, it is removed now.

File: torch/_inductor/fx_passes/freezing_patterns.py (hunk 1 above)
Comment (XiaobingSuper) [reply torch/_inductor/fx_passes/freezing_patterns.py]: yes, updated.

File: torch/_inductor/fx_passes/freezing_patterns.py [hunk 3]
```diff
@@ -118,3 +124,132 @@ def addmm_fuse_replacement_second(inp, w1, w2, w3, b1, b2, b3):
...
+                new_bias = graph.create_node(
+                    "call_function", binary_node.target, (bias, other_reshape)
+                )
+            else:
+                new_bias = graph.create_node(
+                    "call_function", binary_node.target, (0, other_reshape)
+                )
+            conv_args[2] = new_bias
+        else:
+            weight_broad_shape = [1 for _ in range(len(weight_meta_value.shape))]
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply torch/_inductor/fx_passes/freezing_patterns.py]: updated.

**Review by jgong5 — APPROVED** [review]

**Comment by XiaobingSuper** [comment]
@eellison, I wonder if we may remove the conv_bn folding pass by directly calling many times of binary folding to avoid re-tracing the aot_model using ```make_fx```?

**Review by eellison — COMMENTED** [review]
still need to take a look at this but fyi theres a similar pass in jit, maybe worth checking that all of the same conditions are checked.

Can we reuse these tests ? 

https://github.com/pytorch/pytorch/commit/6971149326932586e479b8e983d24dc0d4535999#diff-533101d7d55b7513c4966fcda7f2554d9a7e19b82d54c6b00c6db59686d0e18cR1428-R1510

**Comment by eellison** [comment]
> @eellison, I wonder if we may remove the conv_bn folding pass by directly calling many times of binary folding to avoid re-tracing the aot_model using make_fx?

Yea sounds good !

File: torch/_inductor/fx_passes/freezing_patterns.py [hunk 4]
```diff
@@ -118,3 +123,141 @@ def addmm_fuse_replacement_second(inp, w1, w2, w3, b1, b2, b3):
...
+
+
+@functools.lru_cache(None)
+def binary_folding_init():
+    _conv_args = [Arg() for _ in range(9)]
+    _computation_ops = [aten.convolution.default]
+    _binary_ops = [aten.add.Tensor, aten.sub.Tensor, aten.mul.Tensor, aten.div.Tensor]
+    _computation_calls = [CallFunction(aten.convolution.default, *_conv_args)]
+
+    def _is_constant_node(node):
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply torch/_inductor/fx_passes/freezing_patterns.py:135]: Yes, the checks are following it.

[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/fx_passes/freezing_patterns.py:135]: Cool! Would you mind copying over some of the comments and reusing the tests ? And maybe moving to a separate file ?

I think it'll cool to remove the batch norm pass and replace it with this.

**Comment by jansel** [comment]
Will let @eellison review this one.

File: torch/_inductor/fx_passes/freezing_patterns.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply torch/_inductor/fx_passes/freezing_patterns.py]: Now, copy some of the comments and reuse the tests, for scalar case, I don't support it now, because it needs to use ```aten.full`` to create a new tensor, but it can't be folded(https://github.com/pytorch/pytorch/blob/main/torch/_inductor/freezing.py#L105). I will create another PR to remove the batch norm pass.

**Review by eellison — APPROVED** [review]
Cool, looks good ! 

I guess in a follow up we can remove the batch norm folding and completely replace with this.

File: test/inductor/test_binary_folding.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+            if module == nn.Conv2d:
+                inps.append(inps[-1])
+            if module == nn.Conv3d:
+                inps.append(inps[-1])
+                inps.append(inps[-1])
+
+            inp = torch.rand(inps).to(self.device)
+            out_eager = mod_eager(inp)
+            out_optimized = out_optimized(inp)
+            self.assertEqual(out_optimized, out_eager, atol=2e-04, rtol=1e-5)
```
Comment (XiaobingSuper) [inline test/inductor/test_binary_folding.py:104]: CPU path works fine even using default tolerance, but for the GPU path, there seems has a big variance, but my local machine(A10) can't reproduce it, I need to find another machine to reproduce it.

Comment (XiaobingSuper) [reply test/inductor/test_binary_folding.py]: Skip the pass which cudnn is enabled, it can be passed when disable cudnn convolution.
